File: 01036. Escape a Large Maze/solution.js
/**
 * Problem: 1036. Escape a Large Maze
 *
 * Difficulty: Hard
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 25 ms (Beats 100%)
 */

/**
 * Checks if escape from source to target is possible in a grid with blocked cells
 *
 * @param {number[][]} blocked - Blocked cell coordinates [row, col]
 * @param {number[]} source - Starting position [row, col]
 * @param {number[]} target - Target position [row, col]
 *
 * @returns {boolean} - True if path exists, false otherwise
 */
const isEscapePossible = (blocked, source, target) => {
  // Define DFS function to explore reachable cells from a starting position
  const dfs = (r, c) => {
    // Stop if target found or explored enough cells to confirm no enclosure
    if (flag || seen.size === tgt) return
    // Stop if coordinates are out of bounds
    if (r < 0 || r === n || c < 0 || c === n) return
    // Check if current position matches target coordinates
    if (r === rr && c === cc) return (flag = true)

    // Create unique key for current position using row*n + col formula
    const key = n * r + c

    // Skip if cell already visited or blocked
    if (seen.has(key) || set.has(key)) return

    // Mark current cell as visited
    seen.add(key)

    // Recursively explore all 4 directions (up, down, left, right)
    dfs(r - 1, c)
    dfs(r + 1, c)
    dfs(r, c - 1)
    dfs(r, c + 1)
  }

  // Initialize blocked cells set and grid dimension
  const set = new Set(),
    n = 1000000

  // Convert blocked coordinates to unique keys and store in set
  blocked.forEach((e) => set.add(n * e[0] + e[1]))

  // Initialize tracking variables for DFS exploration
  let seen = new Set(),
    rr = target[0], // Target row coordinate
    flag // Flag to indicate if target is reachable

  // Calculate maximum cells that could be enclosed by blocked cells
  const tgt = (blocked.length ** 2) >> 1,
    cc = target[1] // Target column coordinate

  // Start DFS from source position
  dfs(...source)

  // If target was reached directly, return true
  if (flag) return true

  // If explored cells less than enclosure threshold, source is trapped
  if (seen.size < tgt)
    return false

    // Reset for second DFS from target, clear target coordinates
  ;(seen = new Set()), (rr = null)

  // Start DFS from target position
  dfs(...target)

  // Return true if target can explore enough cells (not trapped)
  return seen.size === tgt
}

File: 01036. Escape a Large Maze/solution.go
/**
 * Problem: 1036. Escape a Large Maze
 *
 * Difficulty: Hard
 *
 * Language: Golang
 *
 * Performance: Runtime - 70 ms (Beats 100%)
 */

func isEscapePossible(blocked [][]int, source []int, target []int) bool {
		// Convert source and target to [2]int for easier map operations
		s := [2]int{source[0], source[1]}
		t := [2]int{target[0], target[1]}
		
		// Grid size limit (10^6)
		m := 1000000

		// Store number of blocked cells for area calculation
		numBlocks := len(blocked)
		
		// Convert blocked cells to map for O(1) lookup
		blocks := map[[2]int]bool{}
		for _, b := range blocked {
			blocks[[2]int{b[0], b[1]}] = true
		}
		
		// Direction vectors: up, left, right, down
		dirs := [4][2]int{{-1, 0}, {0, -1}, {0, 1}, {1, 0}}

		// BFS function to check if we can escape from start to target
		var bfs func(start, target [2]int) bool
		bfs = func(start, target [2]int) bool {
			// Initialize BFS queue with starting position
			q := [][2]int{start}
			
			// Track visited area size
			area := 0
			
			// Keep track of visited cells to avoid cycles
			visited := map[[2]int]bool{start: true}

			// Continue BFS until queue is empty
			for len(q) > 0 {
				// Process current level
				n := len(q)
				for i := 0; i < n; i++ {
					pos := q[i]
					
					// If we reached target or explored enough area to escape any possible enclosure
					// Maximum area that can be enclosed by n blocks is n*(n-1)/2
					if pos == target || area > numBlocks*(numBlocks-1)/2 {
						return true
					}
					
					// Increment visited area
					area++
					
					// Explore all 4 directions
					for _, dir := range dirs {
						x, y := pos[0]+dir[0], pos[1]+dir[1]
						
						// Check if next position is within grid bounds
						if x < 0 || y < 0 || x >= m || y >= m {
							continue
						}
						
						next := [2]int{x, y}
						
						// Skip if position is blocked or already visited
						if blocks[next] || visited[next] {
							continue
						}
						
						// Mark as visited and add to queue
						visited[next] = true
						q = append(q, next)
					}
				}
				// Remove processed level from queue
				q = q[n:]
			}
			
			// Could not reach target or escape the enclosure
			return false
		}

		// Check if both source can reach target AND target can reach source
		// This ensures bidirectional reachability
		return bfs(s, t) && bfs(t, s)
}
